### ISSUE TITLE
Change Request: Support async formatter

### ISSUE THREAD

**Issue by fengzilong** [eslint/eslint#15242]
### ESLint version

8.1.0

### What problem do you want to solve?

Support async formatter

![image](https://user-images.githubusercontent.com/9125255/139874231-870511c8-68cd-42ef-9706-0e13bb3dbebd.png)

[fengzilong/eslint-formatter-mo](https://github.com/fengzilong/eslint-formatter-mo)

I had wrote an eslint formatter with code highlight feature

But it seems the highlighter didn't highlight all tokens well

Recently I want to switch highlighter to [shiki](https://github.com/shikijs/shiki) for a better highlighting result(maybe, I'm trying), but `shiki` has a async highlighter, which is not supported by ESLint currently

### What do you think is the correct solution?

`formatter` can be async function

### Participation

- [X] I am willing to submit a pull request for this change.

**Comment by nzakas** [comment]
Seems like an easy enough change now that most CLI methods are async. I’m in favor. Let’s see what the team thinks.

**Comment by mdjermanovic** [comment]
No objections from me, but should we treat this as a semver-minor or semver-major change? It can be argued that this changes the interface of [`eslint.loadFormatter`](https://eslint.org/docs/developer-guide/nodejs-api#-eslintloadformatternameorpath) because it changes its return type ([Formatter](https://eslint.org/docs/developer-guide/nodejs-api#-formatter-type)).

**Comment by nzakas** [comment]
I think semver-minor, as it doesn’t affect the ability to use existing formatters and doesn’t require any users to do anything differently.

**Comment by mdjermanovic** [comment]
It certainly doesn't affect anything that can be done now, but integrations that use `eslint.loadFormatter()` might be seen as incompatible after this change. For example, `eslint-webpack-plugin` allows users to define `formatter` field in the configuration and loads them with `eslint.loadFormatter()`. We release this in ESLint v8.2.0, and there's a published async formatter. `eslint-webpack-plugin` didn't expect async formatters, so it becomes in a certain way incompatible with ESLint v8.2.0 because it doesn't work with v8.2.0-compatible formatters.

**Comment by nzakas** [comment]
That still doesn’t seem like a semver-major issue to me. eslint.loadFormatter() doesn’t guarantee the signature of the module it is loading. you could pass in any random file and only when you go to call the function will you know if it works (or if, for example, the module returned an object instead of a function)

**Comment by mdjermanovic** [comment]
> That still doesn’t seem like a semver-major issue to me. eslint.loadFormatter() doesn’t guarantee the signature of the module it is loading. you could pass in any random file and only when you go to call the function will you know if it works (or if, for example, the module returned an object instead of a function)

That's true, we cannot really verify and guarantee the signature of the loaded function, and user can always specify a module that isn't actually an ESLint formatter. But, that's then an invalid formatter, and thus user's fault.

A difference here is that integrations will be getting _valid_ ESLint formatters with unexpected signatures, which is a kind of compatibility issue. I don't have a very strong opinion on this, though. If you don't think this is a breaking change for integrations, I'll agree.

**Comment by nzakas** [comment]
I think we can mitigate some of the issues by recommending that integrators always “await” the formatter call so it can work either way.

@btmills can you chime in as well?

**Comment by nzakas** [comment]
Ping @btmills

**Comment by fengzilong** [comment]
If there is an ESLint integration list, I can help send PRs to these repositories to add `await`

**Comment by btmills** [comment]
From a strict typing perspective, going from this signature:

```ts
interface Formatter {
    format(results: LintResult[]): string;
}

class ESLint {
    async loadFormatter(name: string = "stylish"): Promise<Formatter> { /* ... */ }
    // ...
}
```

to this wider return type:

```diff
 interface Formatter {
-    format(results: LintResult[]): string;
+    format(results: LintResult[]): string | Promise<string>;
 }
```

would be a breaking change. But like you all, I'd really like to be able to do this without waiting a year for ESLint v9. I see the argument that it's really the user who's in charge of the return type when they select a formatter. I'd be okay making this change and asking async formatters to document that they may not be compatible with all integrations.

**Comment by nzakas** [comment]
@fengzilong we don’t have an integrations list, which is part of the challenge here.

It sounds like we now have agreement that this change doesn’t have to wait for a major release, so we can proceed.

**Comment by fengzilong** [comment]
> It sounds like we now have agreement that this change doesn’t have to wait for a major release, so we can proceed.

That's great that it can be released soon 🎉
